### DISCUSSION TITLE
1942. The Number of the Smallest Unoccupied Chair

### DISCUSSION THREAD

**Discussion by mah-shamim** [mah-shamim/leet-code-in-php#691] [Q&A] [answered]
**Topics:** `Array`, `Hash Table`, `Heap (Priority Queue)`

There is a party where `n` friends numbered from `0` to `n - 1` are attending. There is an **infinite** number of chairs in this party that are numbered from `0` to `infinity`. When a friend arrives at the party, they sit on the unoccupied chair with the `smallest number`.

- For example, if chairs `0`, `1`, and `5` are occupied when a friend comes, they will sit on chair number `2`.

When a friend leaves the party, their chair becomes unoccupied at the moment they leave. If another friend arrives at that same moment, they can sit in that chair.

You are given a **0-indexed** 2D integer array `times` where <code>times[i] = [arrival<sub>i</sub>, leaving<sub>i</sub>]</code>, indicating the arrival and leaving times of the <code>i<sup>th</sup></code> friend respectively, and an integer `targetFriend`. All arrival times are **distinct**.

Return _the **chair number** that the friend numbered `targetFriend` will sit on_.

**Example 1:**

- **Input:** times = [[1,4],[2,3],[4,6]], targetFriend = 1
- **Output:** 1
- **Explanation:**
  - Friend 0 arrives at time 1 and sits on chair 0.
  - Friend 1 arrives at time 2 and sits on chair 1.
  - Friend 1 leaves at time 3 and chair 1 becomes empty.
  - Friend 0 leaves at time 4 and chair 0 becomes empty.
  - Friend 2 arrives at time 4 and sits on chair 0.\
    Since friend 1 sat on chair 1, we return 1.

**Example 2:**

- **Input:** times = [[3,10],[1,5],[2,6]], targetFriend = 0
- **Output:** 2
- **Explanation:**
  - Friend 1 arrives at time 1 and sits on chair 0.
  - Friend 2 arrives at time 2 and sits on chair 1.
  - Friend 0 arrives at time 3 and sits on chair 2.
  - Friend 1 leaves at time 5 and chair 0 becomes empty.
  - Friend 2 leaves at time 6 and chair 1 becomes empty.
  - Friend 0 leaves at time 10 and chair 2 becomes empty.
    Since friend 0 sat on chair 2, we return 2.

**Example 3:**

- **Input:**  times = [[33889,98676],[80071,89737],[44118,52565],[52992,84310],[78492,88209],[21695,67063],[84622,95452],[98048,98856],[98411,99433],[55333,56548],[65375,88566],[55011,62821],[48548,48656],[87396,94825],[55273,81868],[75629,91467]],   targetFriend = 6
- **Output:** 2

**Example 4:**

- **Input:**  times = [[4435,4436],[2958,2959],[743,744],[3439,3440],[5872,5873],[2507,2508],[3809,3810],[5084,5085],[6509,6510],[7740,7741],[1189,1190],[1277,1278],[7655,7656],[3229,3230],[4527,4528],[2242,2243],[1933,1934],[3348,3349],[6063,6064],[5344,5345],[5708,5709],[4439,4440],[4596,4597],[7572,7573],[330,331],[5153,5154],[2777,2778],[2127,2128],[7024,7025],[6983,6984],[2313,2314],[5961,5962],[7141,7142],[6150,6151],[1011,1012],[2146,2147],[1367,1368],[6741,6742],[64,65],[841,842],[6973,6974],[5002,5003],[2029,2030],[6131,6132],[947,948],[81,82],[2331,2332],[3740,3741],[1486,1487],[6176,6177],[3000,3001],[2805,2806],[6278,6279],[5273,5274],[660,661],[4833,4834],[8293,8294],[7490,7491],[2628,2629],[4032,4033],[5403,5404],[8062,8063],[5043,5044],[6038,6039],[3977,3978],[7045,7046],[1161,1162],[63,64],[2013,2014],[2400,2401],[5947,5948],[4563,4564],[6593,6594],[114,115],[2566,2567],[5297,5298],[8069,8070],[1675,1676],[1746,1747],[7713,7714],[4814,4815],[7651,7652],[6304,6305],[5638,5639],[5881,5882],[1761,1762],[3250,3251],[4391,4392],[1800,1801],[1136,1137],[6489,6490],[556,557],[3729,3730],[7901,7902],[7866,7867],[8312,8313],[1785,1786],[6200,6201],[4469,4470],[5619,5620],[4732,4733],[8039,8040],[4670,4671],[4558,4559],[8076,8077],[3743,3744],[4839,4840],[418,419],[3764,3765],[1118,1119],[3442,3443],[4605,4606],[6806,6807],[179,180],[7753,7754],[1321,1322],[6133,6134],[2456,2457],[6385,6386],[1584,1585],[6111,6112],[5156,5157],[2930,2931],[927,928],[710,711],[501,502],[2308,2309],[1971,1972],[5560,5561],[6216,6217],[5779,5780],[7486,7487],[5750,5751],[6462,6463],[2796,2797],[1543,1544],[6411,6412],[831,832],[5540,5541],[5981,5982],[3925,3926],[199,200],[8332,8333],[7338,7339],[5528,5529],[4962,4963],[3102,3103],[2063,2064],[255,256],[4317,4318],[2207,2208],[1996,1997],[1409,1410],[5095,5096],[7212,7213],[674,675],[2321,2322],[2571,2572],[4678,4679],[5542,5543],[6126,6127],[5270,5271],[6373,6374],[5856,5857],[353,354],[4295,4296],[8216,8217],[6102,6103],[7374,7375],[1578,1579],[3779,3780],[1015,1016],[5278,5279],[2320,2321],[6854,6855],[1722,1723],[1706,1707],[6310,6311],[3080,3081],[3520,3521],[6321,6322],[4505,4506],[4145,4146],[3026,3027],[5105,5106],[3284,3285],[1817,1818],[4374,4375],[4891,4892],[1640,1641],[5057,5058],[1152,1153],[7393,7394],[745,746],[8204,8205],[8321,8322],[7775,7776],[5796,5797],[2626,2627],[3991,3992],[3565,3566],[3671,3672],[6487,6488],[1592,1593],[1144,1145],[5966,5967],[1443,1444],[8224,8225],[1803,1804],[2046,2047],[3586,3587],[1162,1163],[3201,3202],[3783,3784],[6890,6891],[1695,1696],[155,156],[7390,7391],[8173,8174],[5515,5516],[3704,3705],[5422,5423],[4458,4459],[6899,6900],[2169,2170],[1555,1556],[1370,1371],[6224,6225],[7695,7696],[2369,2370],[1560,1561],[1445,1446],[2987,2988],[8114,8115],[673,674],[111,112],[5395,5396],[7122,7123],[1178,1179],[6485,6486],[5101,5102],[62,63],[1345,1346],[3386,3387],[2851,2852],[4294,4295],[5702,5703],[256,257],[7196,7197],.........], targetFriend = 4519
- **Output:** 0

**Example 5:**

- **Input:**  times = [[4,5],[12,13],[5,6],[1,2],[8,9],[9,10],[6,7],[3,4],[7,8],[13,14],[15,16],[14,15],[10,11],[11,12],[2,3],[16,17]],  targetFriend = 15
- **Output:** 0

**Example 6:**

- **Input:** times = [[48210,85285],[46524,48493],[69982,84568],[68053,76419],[93591,97693],[10826,92490],[19366,83205],[90379,92055],[83953,88146],[6466,6795],[96743,98347],[29871,31619],[22949,83080],[58306,91003],[63487,81278],[33256,49208],[60964,98563],[17086,25143],[53332,97203],[18856,26826],[12517,27880],[49357,67393],[96514,97453],[43732,82311],[51936,65126],[17825,77373],[7190,77681],[38818,87041],[60711,84473],[3754,66051],[99983,99984],[5814,55830],[3833,39153],[71632,91461],[99252,99712],[15002,35534],[65097,77671],[6286,87838],[7891,8509],[40505,43365],[54073,74776],[89699,91243],[95158,97485],[31149,42770],[63519,87434],[61935,78221],[2380,28991],[79881,81382],[40190,94494],[41862,73438],[74482,76848],[63016,79828],[12951,13848],[92131,95646],[36030,88066],[86638,92405],[97928,99197],[97679,99457],[69204,95987],[95934,99864],[71117,95167],[1698,50492],[10030,82025],[40893,59120],[13482,73281],[12018,66449],[63542,73061],[98886,99681],[54333,57707],[30903,62762],[10177,27610],[32048,80510],[10389,89810],[62864,81389],[81477,82859],[84637,90371],[15071,27615],[42528,93941],[9344,92289],[54720,88699],[92290,95965],[20387,98404],[17024,86124],[47573,66022],[85502,86949],[52486,61448],[82220,87356],[32475,32637],[58452,59429],[30109,56426],[93029,93594],[51255,59828],[60958,88823],[63538,83957],[276,72672],[28666,54076],[29968,86864],[97360,98067],[1117,34572],[14894,42110],[92657,94778],[5953,72908],[46901,59410],[51178,69717],[1653,79730],[16366,68014],[77911,82792],[81177,82314],[71569,72574],[61956,78298],[75833,88337],[35502,65000],[24047,73173],[9286,24350],[68679,76990],[42891,85895],[65128,83149],[95757,99264],[53956,90056],[75742,77288],[45770,64643],[77996,81439],[21454,59154],[47681,75505],[94171,96038],[36963,96891],[86335,99925],[28351,84927],[90370,91656],[32595,81002],[38745,85563],[82472,99551],[99587,99822],[91156,98592],[8267,86739],[10890,86010],[10411,49492],[10720,74250],[15126,40651],[36192,57999],[69060,98034],[41147,85448],[56870,88507],[58418,77267],[95891,99661],[11388,73213],[40708,50499],[69455,80483],[89904,93067],[5648,27531],[91406,92397],[74965,78521],[48515,87121],[32058,60312],[56742,96316],[44835,54573],[94510,95744],[3323,32938],[6014,53489],[87328,88036],[81596,95508],[8626,99463],[64969,91642],[97634,98280],[84575,99213],[59643,92364],[27550,49499],[28523,30075],[18272,62477],[32870,34885],[93990,99711],[79664,82479],[79885,81238],[43488,96059],[62227,69236],[86015,99363],[52258,87524],[84395,94009],[260,61980],[94639,96103],[81610,93673],[8375,73739],[92648,98303],[43573,80190],[97479,99688],[12786,62429],[85068,99354],[65625,73789],[39586,80043],[60278,99931],[69644,77411],[87222,92178],[52067,89116],[32663,66802],[32590,43620],[11531,27051],[164,63564],[98119,99886],[60542,86235],[65210,80094],[82852,97232],[66702,71196],[39672,50854],[97270,98306],[4351,52670],[93525,95535],[15402,50714],[42024,77025],[46532,59101],[8418,96138],[77412,91249],[4775,11015],[22526,76789],[23669,73864],[64077,92284],[86509,99576],[23456,41142],[45007,65172],[4577,9088],[96650,98345],[35929,70561],[47528,72447],[8645,66485],[18387,31301],[15698,31613],[8899,74243],[94207,98592],[67673,81051],[48106,69025],[39626,94628],[9983,72729],[67724,68340],[29765,59531],[89752,90556],[21787,54950],[5324,34185],[64786,87911],[97565,97961],[81401,85960],[81947,91400],[55608,85874],[30719,95522],[39371,84566],[92294,97705],[96165,99620],[57736,74149],[13000,81640],[52622,95627],[48409,59629],[27390,51248],[63747,69043],[65597,81709],[10095,19419],[15234,47659],[34309,44757],[33863,48810],[91084,92091],[67817,77971],[72221,82727],[16295,25043],[64353,98511],[93077,96137],[16133,30596],[75551,89906],[60231,96814],[41132,88940],[84501,90563],[47674,68829],[55742,92010],[79179,90956],[55505,60411],[7914,76864],[98625,98825],[145,92328],[90200,95685],[70600,82915],[43649,91009],[78906,93751],[97838,98151],[83318,89717],[98298,99752],[73204,77019],[7553,30523],[79612,98738],[72655,80248],[68949,81988],[49684,66571],[40856,45815],[83263,84874],[66452,98833],[9840,86578],[68947,82875],[45036,74480],[8597,45642],[16608,65906],[62817,75808],[45101,93029],[69711,89393],[30250,33181],[73993,88229],[68561,97074],[10416,28408],[18613,91871],[43356,97486],[16478,80062],[97635,98588],[62841,65681],[39104,94753],[89839,96542],[90255,92840],[34444,59080],[36182,69681],[84572,91103],[29941,79497],[36183,68570],[73978,99514],[64217,79142],[98422,99473],[70793,97469],[93645,95936],.........], targetFriend = 8916
- **Output:** 2235

**Example 7:**

- **Input:** times = [[2241,100000],[2852,100000],[5733,100000],[9191,100000],[3272,100000],[2018,100000],[6449,100000],[290,100000],[8719,100000],[4836,100000],[3939,100000],[4903,100000],[1936,100000],[9391,100000],[5910,100000],[1511,100000],[1839,100000],[6922,100000],[392,100000],[9233,100000],[8175,100000],[8520,100000],[5904,100000],[490,100000],[2931,100000],[7972,100000],[7522,100000],[9083,100000],[7653,100000],[955,100000],[135,100000],[4501,100000],[7031,100000],[8180,100000],[6694,100000],[6277,100000],[2061,100000],[6337,100000],[4539,100000],[5685,100000],[700,100000],[7568,100000],[1637,100000],[2600,100000],[2647,100000],[9280,100000],[7895,100000],[3114,100000],[14,100000],[9371,100000],[5615,100000],[8578,100000],[7042,100000],[151,100000],[868,100000],[6735,100000],[6199,100000],[8362,100000],[9615,100000],[9705,100000],[2031,100000],[5189,100000],[5767,100000],[4094,100000],[957,100000],[5939,100000],[9193,100000],[9383,100000],[2528,100000],[9222,100000],[676,100000],[5885,100000],[9888,100000],[2452,100000],[7355,100000],[6998,100000],[4061,100000],[5129,100000],[4364,100000],[9348,100000],[3454,100000],[6239,100000],[3491,100000],[968,100000],[3724,100000],[1629,100000],[6180,100000],[8219,100000],[1842,100000],[673,100000],[8101,100000],[2014,100000],[7500,100000],[8231,100000],[8566,100000],[6216,100000],[7999,100000],[6397,100000],[275,100000],[843,100000],[1471,100000],[1627,100000],[1575,100000],[4032,100000],[829,100000],[2601,100000],[3652,100000],[6807,100000],[9286,100000],[6693,100000],[5684,100000],[4125,100000],[3503,100000],[2727,100000],[8897,100000],[3924,100000],[7961,100000],[3027,100000],[655,100000],[7294,100000],[5547,100000],[2790,100000],[6543,100000],[1724,100000],[6478,100000],[7992,100000],[1851,100000],[1779,100000],[4850,100000],[4598,100000],[1948,100000],[8208,100000],[7056,100000],[2313,100000],[8648,100000],[6276,100000],[1156,100000],[6952,100000],[685,100000],[4522,100000],[2365,100000],[4440,100000],[4940,100000],[3322,100000],[1560,100000],[7682,100000],[9425,100000],[6733,100000],[8398,100000],[5025,100000],[543,100000],[178,100000],[3792,100000],[23,100000],[299,100000],[6755,100000],[7153,100000],[7679,100000],[7715,100000],[394,100000],[2213,100000],[8811,100000],[541,100000],[6684,100000],[8552,100000],[9413,100000],[2097,100000],[16,100000],[5612,100000],[469,100000],[4193,100000],[2605,100000],[2692,100000],[8710,100000],[4269,100000],[7520,100000],[9226,100000],[2765,100000],[6833,100000],[9064,100000],[1036,100000],[6245,100000],[8667,100000],[5518,100000],[7586,100000],[4968,100000],[6439,100000],[6928,100000],[5318,100000],[8943,100000],[8882,100000],[4992,100000],[1204,100000],[2371,100000],[3432,100000],[9100,100000],[4035,100000],[9497,100000],[4970,100000],[5446,100000],[6687,100000],[28,100000],[1769,100000],[4211,100000],[8482,100000],[860,100000],[3896,100000],[9927,100000],[7221,100000],[1939,100000],[9783,100000],[2377,100000],[5406,100000],[9991,100000],[5801,100000],[7110,100000],[8744,100000],[1016,100000],[7290,100000],[7975,100000],[680,100000],[9576,100000],[691,100000],[1477,100000],[6296,100000],[3850,100000],[4510,100000],[7574,100000],[5175,100000],[3293,100000],[1127,100000],[4772,100000],[4951,100000],[2063,100000],[4716,100000],[2363,100000],[7404,100000],[3453,100000],[5437,100000],[1187,100000],[6026,100000],[5535,100000],[6125,100000],[2830,100000],[1031,100000],[6573,100000],[7825,100000],[4958,100000],[5841,100000],[4442,100000],[7218,100000],[2899,100000],[1848,100000],[1923,100000],[8638,100000],[2283,100000],[8100,100000],[2348,100000],[3549,100000],[3838,100000],[1573,100000],[6472,100000],[3008,100000],[8523,100000],[8144,100000],[4514,100000],[3758,100000],[1880,100000],[2379,100000],[1017,100000],[8002,100000],[8080,100000],[5264,100000],[6067,100000],[695,100000],[6958,100000],[431,100000],[6141,100000],.........],  targetFriend = 2145
- **Output:** 9998

**Example 8:**

- **Input:** times = [[22426,84367],[75329,93393],[16727,83539],[36174,69610],[1205,96372],[92090,92614],[54318,68966],[2108,45037],[33071,97071],[10136,32984],[61156,98710],[5352,26672],[69124,73060],[36589,67343],[9489,99078],[56391,79127],[88275,91841],[47717,71477],[77283,94145],[34279,62828],[93594,94673],[86554,86912],[39556,46739],[53675,75259],[49687,76186],[98841,99920],[71395,71859],[77560,90201],[19800,27407],[29293,36331],[19298,69257],[45777,51071],[88102,91436],[13402,91278],[75702,90696],[35933,80138],[5409,76632],[69427,76986],[22387,42315],[81578,87718],[23495,89502],[31443,58262],[78751,84610],[49378,73936],[57481,99920],[63048,83860],[29535,72348],[89222,92226],[91976,96524],[93321,95260],[2985,80101],[33096,45591],[34892,89098],[71393,75635],[71885,97563],[74351,80224],[70755,97651],[31640,57431],[85098,97725],[52658,58804],[45399,66232],[26851,39059],[81438,83643],[12196,47170],[3859,67920],[61379,85651],[28430,95048],[92417,94207],[93228,96597],[6785,41005],[57995,89468],[89198,89557],[83057,86667],[89833,89939],[44572,86692],[7582,67737],[23988,84070],[67055,98564],[17603,19265],[47502,78080],[7604,31781],[39812,71979],[92551,99714],[61947,79505],[24114,49424],[85626,87737],[58459,62467],[82399,98463],[29555,95257],[29806,71224],[53662,88132],[2228,95288],[14511,93026],[81984,85276],[25902,99856],[20185,29203],[98840,99213],[14791,21666],[26187,45536],[3948,23037],[46754,61956],[53204,97607],[73600,91312],[53148,74118],[8123,75415],[72830,81231],[71516,76002],[48913,61144],[53494,94575],[11622,30017],[89469,98388],[86631,90355],[70085,82969],[66542,82854],[53223,76940],[95413,98935],[84331,92617],[59795,96805],[62877,77297],[90834,92357],[98516,99205],[1335,30749],[36134,39169],[10902,15147],[85022,88653],[96363,97541],[46187,84429],[47101,91808],[6381,29128],[90360,93219],[678,14919],[31908,60036],[40179,82797],[67855,74554],[80220,95471],[98962,99729],[3438,37747],[64847,78496],[2696,41421],[31478,31717],[52990,54936],[98901,99063],[29069,43478],[63633,80235],[80021,80679],[48405,62811],[86712,92198],[11335,69289],[63829,96943],[88025,99197],[22243,78723],[39726,47031],[3674,76470],[28839,85178],[49939,98708],[70724,79599],[56434,88892],[46924,78658],[88332,98027],[149,51884],[31693,83801],[2982,8380],[53085,88474],[99479,99603],[74758,96516],[72069,88990],[89990,93045],[62167,84228],[12199,29553],[49976,56856],[67653,98565],[68522,76794],[67032,91388],[81732,92415],[6704,14173],[3813,57358],[30028,82717],[39697,52527],[44618,45237],[74779,96499],[87531,96157],[38515,90279],[57110,66630],[68505,70468],[77731,80244],[81590,99464],[72188,80823],[89076,94703],[10215,70391],[57010,61892],[95915,96351],[30708,80862],[85150,87096],[42273,99632],[17736,65771],[56542,81567],[19812,80923],..........],  targetFriend = 2304
- **Output:** 2027

**Constraints:**

- `n == times.length`
- <code>2 <= n <= 10<sup>4</sup></code>
- `times[i].length == 2`
- <code>1 <= arrivali < leavingi <= 10<sup>5</sup></code>
- `0 <= targetFriend <= n - 1`
- Each <code>arrival<sub>i</sub></code> time is **distinct**.

**Hint:**
1. Sort times by arrival time.
2. for each `arrival_i` find the smallest unoccupied chair and mark it as occupied until `leaving_i`.

**Comment by kovatz** [comment, marked as answer]
We need to simulate the arrival and departure of friends at a party and manage the assignment of the smallest available chairs. Here's the step-by-step approach:

### Approach:

1. **Sort by Arrival Times**:
   - Sort the `times` array based on the arrival time.
   - This helps process friends in the order of their arrivals.

2. **Use Priority Queues (Min-Heaps)**:
   - Use a min-heap to manage unoccupied chairs (`availableChairs`).
   - Use another min-heap to track the occupied chairs along with their release times (`occupiedChairs`).

3. **Iterate Through Sorted Times**:
   - For each friend's arrival:
      - Free up chairs of friends who have already left by comparing the current time to the friends' departure times. Move those chairs back to `availableChairs`.
      - Assign the smallest available chair (from `availableChairs`) to the arriving friend. If no chairs are free, assign the next smallest unused chair.
      - Push this chair and the friend's leaving time into `occupiedChairs`.

4. **Track Target Friend's Chair**:
   - When the `targetFriend` arrives, record the chair assigned to them.

5. **Edge Cases**:
   - Consider cases where multiple friends arrive at the same time.
   - Ensure that the smallest chair is always assigned.

Let's implement this solution in PHP: **[1942. The Number of the Smallest Unoccupied Chair](https://github.com/mah-shamim/leet-code-in-php/tree/main/algorithms/001942-the-number-of-the-smallest-unoccupied-chair/solution.php)**

```php
<?php
/**
 * @param Integer[][] $times
 * @param Integer $targetFriend
 * @return Integer
 */
function smallestChair($times, $targetFriend) {
    $nextUnsatChair = 0;
    $emptyChairs = new SplMinHeap(); // Min heap for empty chairs
    $occupied = new SplMinHeap(); // Min heap for occupied chairs

    // Append the index to each time entry
    foreach ($times as $i => $time) {
        $times[$i][] = $i; // Add index to the end of each time entry
    }

    // Sort times based on arrival time
    usort($times, function($a, $b) {
        return $a[0] - $b[0];
    });

    foreach ($times as $time) {
        $arrival = $time[0];
        $leaving = $time[1];
        $i = $time[2];

        // Free chairs that have become available
        while (!$occupied->isEmpty() && $occupied->top()[0] <= $arrival) {
            $emptyChairs->insert($occupied->top()[1]);
            $occupied->extract(); // Remove the top element
        }

        // Check if this is the target friend
        if ($i == $targetFriend) {
            return $emptyChairs->isEmpty() ? $nextUnsatChair : $emptyChairs->top();
        }

        // Allocate chair
        if ($emptyChairs->isEmpty()) {
            $occupied->insert([$leaving, $nextUnsatChair++]);
        } else {
            $occupied->insert([$leaving, $emptyChairs->top()]);
            $emptyChairs->extract(); // Remove the used chair
        }
    }

    return -1; // This should never happen
}

// Example usage:
$cost1 = [[15, 96], [36, 2]];
$cost2 = [[1, 3, 5], [4, 1, 1], [1, 5, 3]];
$cost3 = [[2, 5, 1], [3, 4, 7], [8, 1, 2], [6, 2, 4], [3, 8, 8]];

// Example usage:
$times1 = [[1, 4], [2, 3], [4, 6]];
$targetFriend1 = 1;
echo smallestChair($times1, $targetFriend1);  // Output: 1

$times2 = [[3, 10], [1, 5], [2, 6]];
$targetFriend2 = 0;
echo smallestChair($times2, $targetFriend2);  // Output: 2
?>
```

### Explanation:

- **Sorting** ensures friends are processed in order of their arrival times.
- **Using a min-heap for available chairs** allows us to always assign the smallest unoccupied chair.
- **Occupied chairs heap** helps track when friends leave, making their chairs available again.
- When the `targetFriend` arrives, we directly return the chair assigned to them.

### Complexity:

- **Time Complexity**: _**O(n log n)**_, primarily due to sorting the `times` array and managing the heaps.
- **Space Complexity**: _**O(n)**_, for storing chairs in the heaps.

This solution efficiently handles the assignment of chairs, ensuring that each friend gets the smallest available chair, and accurately tracks when chairs become available again.

**Reply by mah-shamim** [reply]
Thanks to solve the "Number of the Smallest Unoccupied Chair" problem

**Reply by kovatz** [reply]
Thanks
